Silence errors in expressions caused by bare traits in paths in 2021 edition

estebank opened the pull request:
Address #51077 in editions>=2021 by treating `<Trait>` as a `<{type error}>` instead of `<dyn Trait>`, silencing subsequent errors. We now emit a single error ("missing `dyn`").

```
error[E0782]: trait objects must include the `dyn` keyword
 --> f800.rs:2:15
  |
2 | let x: u32 = <Default>::default();
  |               ^^^^^^^
  |
help: add `dyn` keyword before this trait
  |
2 | let x: u32 = <dyn Default>::default();
  |               +++
```

instead of 6

```
error[E0782]: trait objects must include the `dyn` keyword
 --> f800.rs:2:15
  |
2 | let x: u32 = <Default>::default();
  |               ^^^^^^^
  |
help: add `dyn` keyword before this trait
  |
2 | let x: u32 = <dyn Default>::default();
  |               +++

error[E0038]: the trait `Default` cannot be made into an object
 --> f800.rs:2:15
  |
2 | let x: u32 = <Default>::default();
  |               ^^^^^^^ `Default` cannot be made into an object
  |
  = note: the trait cannot be made into an object because it requires `Self: Sized`
  = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>

error[E0277]: the size for values of type `dyn Default` cannot be known at compilation time
   --> f800.rs:2:15
    |
2   | let x: u32 = <Default>::default();
    |               ^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `dyn Default`
note: required by a bound in `default`
   --> /home/gh-estebank/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/default.rs:104:20
    |
104 | pub trait Default: Sized {
    |                    ^^^^^ required by this bound in `Default::default`
...
136 |     fn default() -> Self;
    |        ------- required by a bound in this associated function

error[E0308]: mismatched types
 --> f800.rs:2:14
  |
2 | let x: u32 = <Default>::default();
  |        ---   ^^^^^^^^^^^^^^^^^^^^ expected `u32`, found `dyn Default`
  |        |
  |        expected due to this
  |
  = note:      expected type `u32`
          found trait object `dyn Default`
  = help: `u32` implements `Default` so you could change the expected type to `Box<dyn Default>`

error[E0038]: the trait `Default` cannot be made into an object
 --> f800.rs:2:14
  |
2 | let x: u32 = <Default>::default();
  |              ^^^^^^^^^^^^^^^^^^^^ `Default` cannot be made into an object
  |
  = note: the trait cannot be made into an object because it requires `Self: Sized`
  = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>

error[E0277]: the size for values of type `dyn Default` cannot be known at compilation time
 --> f800.rs:2:14
  |
2 | let x: u32 = <Default>::default();
  |              ^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `dyn Default`
  = note: the return type of a function must have a statically known size
```

In all editions, whenever we encounter `!Sized` E0277 *and* object safety E0038 errors for the same expression, silence the former:

```
warning: trait objects without an explicit `dyn` are deprecated
  --> $DIR/trait-missing-dyn-in-qualified-path.rs:5:19
   |
LL |     let x: u32 = <Default>::default();
   |                   ^^^^^^^
   |
   = warning: this is accepted in the current edition (Rust 2018) but is a hard error in Rust 2021!
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
   = note: `#[warn(bare_trait_objects)]` on by default
help: if this is an object-safe trait, use `dyn`
   |
LL |     let x: u32 = <dyn Default>::default();
   |                   +++

error[E0038]: the trait `Default` cannot be made into an object
  --> $DIR/trait-missing-dyn-in-qualified-path.rs:5:19
   |
LL |     let x: u32 = <Default>::default();
   |                   ^^^^^^^ `Default` cannot be made into an object
   |
   = note: the trait cannot be made into an object because it requires `Self: Sized`
   = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>

error[E0308]: mismatched types
  --> $DIR/trait-missing-dyn-in-qualified-path.rs:5:18
   |
LL |     let x: u32 = <Default>::default();
   |            ---   ^^^^^^^^^^^^^^^^^^^^ expected `u32`, found `dyn Default`
   |            |
   |            expected due to this
   |
   = note:      expected type `u32`
           found trait object `dyn Default`
   = help: `u32` implements `Default` so you could change the expected type to `Box<dyn Default>`

error[E0038]: the trait `Default` cannot be made into an object
  --> $DIR/trait-missing-dyn-in-qualified-path.rs:5:18
   |
LL |     let x: u32 = <Default>::default();
   |                  ^^^^^^^^^^^^^^^^^^^^ `Default` cannot be made into an object
   |
   = note: the trait cannot be made into an object because it requires `Self: Sized`
   = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
```

estebank committed:
Silence errors in experssions caused by bare traits in paths

Address #51077 in editions>=2021 by treating `<Trait>` as a `<{type error}>` instead of `<dyn Trait>`, silencing subsequent errors. We new emit a single error ("missing `dyn`").

```
error[E0782]: trait objects must include the `dyn` keyword
 --> f800.rs:2:15
  |
2 | let x: u32 = <Default>::default();
  |               ^^^^^^^
  |
help: add `dyn` keyword before this trait
  |
2 | let x: u32 = <dyn Default>::default();
  |               +++
```

instead of 6

```
error[E0782]: trait objects must include the `dyn` keyword
 --> f800.rs:2:15
  |
2 | let x: u32 = <Default>::default();
  |               ^^^^^^^
  |
help: add `dyn` keyword before this trait
  |
2 | let x: u32 = <dyn Default>::default();
  |               +++

error[E0038]: the trait `Default` cannot be made into an object
 --> f800.rs:2:15
  |
2 | let x: u32 = <Default>::default();
  |               ^^^^^^^ `Default` cannot be made into an object
  |
  = note: the trait cannot be made into an object because it requires `Self: Sized`
  = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>

error[E0277]: the size for values of type `dyn Default` cannot be known at compilation time
   --> f800.rs:2:15
    |
2   | let x: u32 = <Default>::default();
    |               ^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `dyn Default`
note: required by a bound in `default`
   --> /home/gh-estebank/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/default.rs:104:20
    |
104 | pub trait Default: Sized {
    |                    ^^^^^ required by this bound in `Default::default`
...
136 |     fn default() -> Self;
    |        ------- required by a bound in this associated function

error[E0308]: mismatched types
 --> f800.rs:2:14
  |
2 | let x: u32 = <Default>::default();
  |        ---   ^^^^^^^^^^^^^^^^^^^^ expected `u32`, found `dyn Default`
  |        |
  |        expected due to this
  |
  = note:      expected type `u32`
          found trait object `dyn Default`
  = help: `u32` implements `Default` so you could change the expected type to `Box<dyn Default>`

error[E0038]: the trait `Default` cannot be made into an object
 --> f800.rs:2:14
  |
2 | let x: u32 = <Default>::default();
  |              ^^^^^^^^^^^^^^^^^^^^ `Default` cannot be made into an object
  |
  = note: the trait cannot be made into an object because it requires `Self: Sized`
  = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>

error[E0277]: the size for values of type `dyn Default` cannot be known at compilation time
 --> f800.rs:2:14
  |
2 | let x: u32 = <Default>::default();
  |              ^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `dyn Default`
  = note: the return type of a function must have a statically known size
```

fmease commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:
```diff
@@ -2080,7 +2080,18 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
+                        tcx.parent_hir_node(hir_ty.hir_id)
+                    {
+                        // `tests/ui/suggestions/issue-116434-2021.rs` and others would otherwise
+                        // fail with no error beign emitted. We restrict the silencing only to
```
What do you mean by "fail with no error [being] emitted"? In this branch we have `guard : ErrorGuaranteed`, we should've emitted at least one error?

estebank replied:
For some reason (which I don't understand, as it goes against my understanding of how this is meant to work), the stashed error ends up never getting emitted, which of course causes an ICE for an error being constructed but not emitted. This check does side step that.

fmease replied:
nnethercote has dug into this a while back and written a nice explainer: https://github.com/rust-lang/rust/issues/120482#issuecomment-1937991551

estebank committed:
Silence unsized types errors in some expressions

When encountering an unsized E0277 error on a method that returns `Self` silence that error as it a prior unsized error will have been emitted when constructing the value of `Self`.

On E0227 unsized error caused on a bare trait associated function with object safety violations, silence it.

fmease reviewed (commented):
I love the core change (→ `{type error}`) but I'm not fond of the workaround because it doesn't seem very robust and I bet it doesn't fix all pathological cases without having looked further into it.

Would you be willing to look into fixing the `stash`/`downgrade` issue?

fmease commented:
I wonder if we could just ignore the request to downgrade a stashed diagnostic?

estebank committed:
Avoid potential for ICE by not stashing delayed as bug error and silence more redundant errors

Instead of calling `dowgrade_to_delayed_bug` and then stashing a diagnostic, we either stash the diagnostic or we `delay_as_bug` it.

Remove unnecessary restriction of silencing of knock down errors on bare trait objects in edition >=2021.

estebank commented:
@fmease I see the problem:

 - `delay_as_bug` consumes the diagnostic, in the same way that `emit` does,
 - `downgrade_to_delayed_bug` *doesn't*,
 - this interacts badly with stashing

I changed the code here in hir analysis so that we either stash or delay as bug, but never both. Removed the unnecessary check and now a few more cases get correctly silenced in the test suite.

We could look to better account for downgraded diagnostics that got stashed and not ICE in that case, but for now I think it is not as bad an idea to be very noisy about that happening. That way we can actually make a determination (if annoyingly), instead of paper over it and miss potential unintended mistakes.

estebank committed:
Do not silence some sized errors as they would cause ICEs

estebank committed:
Add more tracking for the HIR source of well-formedness requirement

estebank committed:
Add tests for various `dyn Trait` in path cases

estebank committed:
Add more HIR tracking information to `ObligationCauseCode`

This data will help with extending E0038 object safety errors.

estebank committed:
On object safety errors, point at specific object unsafe trait in path

We use the new HIR node tracking for well-formedness obligations to
point at the specific trait object in a path/type (`dyn T` instead of a
whole `Box<dyn T>`, for example).

```
error[E0038]: the trait `MapLike` cannot be made into an object
  --> $DIR/file.rs:47:16
   |
LL |         as Box<dyn Trait>;
   |                ^^^^^^^^^ `Trait` cannot be made into an object
```

We also provide a structured suggestion for `<dyn Trait>::assoc`:

```
   = help: when writing `<dyn Trait>::function` you are requiring `Trait` be "object safe", which it isn't
help: you might have meant to access the associated function of a specific `impl` to avoid requiring "object safety" from `Trait`, either with some explicit type...
   |
LL |     </* Type */ as Trait>::function();
   |      ~~~~~~~~~~~~~
help: ...or rely on inference if the compiler has enough context to identify the desired type on its own...
   |
LL -     <dyn Trait>::function();
LL +     Trait::function();
   |
help: ...which is equivalent to
   |
LL |     <_ as Trait>::function();
   |      ~~~~
```

estebank committed:
Move `FnCtxt::get_fn_decl` to `TyCtxt`

estebank committed:
Deduplicate more E0038 object safety errors by pointing at type more often

On casts to `dyn Trait` caused by explicit `as` or return type, we point at the type instead of the expression. Silence note with context about the cast when we'd point at the type, which allows automatic deduplication to happen.

estebank commented:
If reviewing all commits is too much (which is likely), feel free to review until the 4th commit and the rest can go on their own PR.

fmease commented:
I guess suggesting `<_ as Default>::default()` or `Default::default()` is the next step here? Esp. if the trait in question is not object-safe (without intro'ing cycle errors ofc)?

fmease commented:
On a second thought, the help message ``add `dyn` keyword before this trait`` is way too assertive. I remember we merged a PR of yours which changed such messages if the trait wasn't object-safe which lead to ICEs and cycle errors and thus we had to partially revert it. We then had a PR by sb. else that changed the wording to sth. like ``add `dyn` keyword before this trait if this trait is object-safe``.

All that too say, while the output on master is horrendously verbose, the output by the current PR might be too terse because it would take the user an additional edit (addition of `dyn`) to learn that wouldn't be correct after all as `Default` is not object-safe.

I guess I'm fine that as long as we/you/I do follow-up work and make the suggestion more correct.

fmease commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:
```diff
@@ -2080,7 +2080,18 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             }
             hir::TyKind::TraitObject(bounds, lifetime, repr) => {
-                self.prohibit_or_lint_bare_trait_object_ty(hir_ty, in_path);
+                if let Some(guard) = self.prohibit_or_lint_bare_trait_object_ty(hir_ty, in_path) {
```
minor: it's `guar` (for *guaranteed*, not the word *guard*) or `reported`.

fmease commented on tests/ui/dyn-keyword/trait-missing-dyn-in-qualified-path.edition2021.stderr:
```diff
@@ -0,0 +1,14 @@
+LL |     let x: u32 = <Default>::default();
+   |                   ^^^^^^^
+   |
+help: add `dyn` keyword before this trait
```
Right, so I guess we should add `if this trait is object-safe` until we check for object-safety again(?)

fmease commented on tests/ui/resolve/issue-111312.rs:
```diff
@@ -7,7 +7,6 @@ trait Has {
-    HasNot::has();
-    //~^ ERROR trait objects must include the `dyn` keyword
-    //~| ERROR no function or associated item named `has` found for trait `HasNot`
+    <dyn HasNot>::has();
```
:/ So does this PR undoes the work done by #111588? I'm pretty sure that #111312 was specifically about using `HasNot::has`. This change is not faithful to the original version.

estebank replied:
I'll see if I can do something in E0782 to at the resolve, but I am generally of the position that if an error leads to another error that has the right suggestion, that is reasonable.

estebank replied:
Now I check if `HasNot` is object safe and if the type is part of a qualified path to an associated item that didn't resolve (unresolved assoc item can be because it doesn't exist in the trait *or* because it exists but an appropriate self ty couldn't be found). If it is, I don't turn it into `{type error}`, which makes the error about "`has` not found" still be emitted. But if the trait is not object safe, we still turn it into a `{type error}` to avoid any further errors. At the same time, I've added more context to the E0782 about object safety so we don't lose any information when emitting only that one.

fmease commented on tests/ui/resolve/issue-111727.rs:
```diff
@@ -3,5 +3,4 @@
 fn main() {
     std::any::Any::create();
     //~^ ERROR trait objects must include the `dyn` keyword
```
This represents another #111312 / #111588 regression.

fmease commented on compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:
```diff
@@ -639,6 +639,32 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                             err.span_label(span, explanation);
                         }
 
+                        if Some(trait_ref.def_id()) == self.tcx.lang_items().sized_trait() {
+                            if !self.tcx.object_safety_violations(trait_ref.def_id()).is_empty() {
```
`Sized` is never object-safe. Why the need to check the `object_safe_violations`?

fmease replied:
Could you maybe add a comment outlining the motivation for delaying the error in these cases for future readers?

estebank replied:
This is now completely gone.

fmease commented on compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:
```diff
@@ -639,6 +639,32 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
+                                    ) => {
+                                        return err.delay_as_bug();
+                                    }
+                                    (ObligationCauseCode::SizedCallReturnType(did), _) => {
```
I'd like to note that these checks are like super specific / not really general. Cost in complexity vs. benefit.

fmease commented on compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:
```diff
@@ -641,30 +641,21 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
-                                        ObligationCauseCode::WhereClauseInExpr(..),
-                                        ty::Dynamic(..),
-                                    ) => {
+                                if let ObligationCauseCode::SizedCallReturnType(did)
```
Right, I've just arrived at this commit (the 4th). Noticing you had to make this change less "aggressive" makes me think that this branch is still super fragile. Like I don't know if the `delay_as_bug` is correct in all perceivable cases

estebank replied:
All `delay_as_bug`s are now gone after the "Fix tidy" commit.

fmease commented on compiler/rustc_hir_analysis/src/check/wfcheck.rs:
(not to self): I haven't properly reviewed commit "Add more tracking for HIR source well-formedness requirement" yet.

fmease commented on tests/ui/dyn-keyword/inferred-type-in-qualified-path-for-with-one-impl.rs:
At commit "Add tests for various `dyn Trait` in path cases" right now. I think we should make sure to remove files that are duplicates of the new tests. For example ``rg 'cannot call associated function on trait without specifying the corresponding `impl` type'  tests -t rust`` gave me `tests/ui/type/issue-101866.rs`, `tests/ui/suggestions/issue-104328.rs`, `tests/ui/suggestions/issue-104327.rs`. We might want to remove them or move them here. I haven't looked at the contents.

fmease replied:
I feel like this specific test doesn't actually fit into `dyn-keyword/`?

fmease reviewed (commented):
Note to self: I've stopped my review right before commit "Add more HIR tracking information to ObligationCauseCode". I see that you're working on sugg'ing `<_ as Default>::default`. I've read your comment about splitting this PR. I don't really care as long as the history is self-explanatory.

You might want to update the PR description as well.

---

Regarding the actual changes I still need to review & re-review things but already have some feedback.

bors commented:
:umbrella: The latest upstream changes (presumably #126108) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank committed:
Move suggestions out of main error logic

estebank committed:
Tweak wording and only suggest `dyn Trait` if `Trait` is object safe

estebank committed:
Suggest `<_ as Default>::default()` when given `<Default>::default()`

estebank committed:
review comment: `guard` -> `guar`

estebank committed:
review comment: unnecessary check for object safety on sized trait

estebank committed:
Add explanatory comment

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9439546079/job/25997797743) [(plain)](https://github.com/rust-lang/rust/commit/5f22c02425acc230457ba2d609df3cbd3e17c0e2/checks/25997797743/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:a5ac7e51b41094c92402da3b24376905380afc29)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.434   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.453 Collecting boolean-py==4.0
#13 0.460   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.478 Collecting chardet==5.1.0
---
#13 3.613 Building wheels for collected packages: reuse
#13 3.614   Building wheel for reuse (pyproject.toml): started
#13 3.938   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.939   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.939   Stored in directory: /tmp/pip-ephem-wheel-cache-taifo5_k/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.942 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.964   Attempting uninstall: setuptools
#13 3.965     Found existing installation: setuptools 59.6.0
#13 3.966     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.966     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.966     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.642 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.642 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.168 Collecting virtualenv
#13 5.225   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.409      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 21.6 MB/s eta 0:00:00
#13 5.462 Collecting platformdirs<5,>=3.9.1
#13 5.469   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.490 Collecting distlib<1,>=0.3.7
#13 5.500   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.550 Collecting filelock<4,>=3.12.2
#13 5.558   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.558   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.643 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.807 Successfully installed distlib-0.3.8 filelock-3.14.0 platformdirs-4.2.2 virtualenv-20.26.2
#13 DONE 5.9s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      198592 kB
DirectMap2M:     8189952 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/503dfcf4e03089cf0c06d10fc69b5477a2f766ef/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-503dfcf4e03089cf0c06d10fc69b5477a2f766ef-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5044 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:652: \
Use a single space after dots in comments.
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.0)
some tidy checks failed
Build completed unsuccessfully in 0:01:08
  local time: Sun Jun  9 21:21:25 UTC 2024
  network time: Sun, 09 Jun 2024 21:21:25 GMT
```

</details>

estebank committed:
Fix tidy

estebank committed:
Remove conditional delay_as_bug from "missing `dyn` error"

estebank committed:
Remove `delay_as_bug` for fn call with unsized `Self` that return `Self`.

estebank commented:
The last two commits remove the two `delay_as_bug` so that you can see the diff. After *all the other* changes, I'm not as married to silencing those errors that are left.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9453614188/job/26039367676) [(plain)](https://github.com/rust-lang/rust/commit/8ee6812c282bb0dd65a460ccc88c0a8b49e40f0c/checks/26039367676/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 31.6s done
#16 DONE 40.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[endgroup]
##[group]Testing stage2 compiletest suite=crashes mode=crashes (x86_64-unknown-linux-gnu)

running 230 tests
............................................................................2024-06-10T18:46:15.667767Z ERROR compiletest::runtest: fatal error, panic: "test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit."
[crashes] tests/crashes/120241-2.rs ... F
[crashes] tests/crashes/120241-2.rs ... F
.2024-06-10T18:46:15.698585Z ERROR compiletest::runtest: fatal error, panic: "test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit."
[crashes] tests/crashes/120241.rs ... F
[crashes] tests/crashes/120241.rs ... F
2024-06-10T18:46:15.698914Z ERROR compiletest::runtest: fatal error, panic: "test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit."
[crashes] tests/crashes/120482.rs ... F
......  88/230
........................................................................................ 176/230
........................................................................................ 176/230
..2024-06-10T18:46:16.415655Z ERROR compiletest::runtest: fatal error, panic: "test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit."
[crashes] tests/crashes/125512.rs ... F
...................................................

failures:
failures:

---- [crashes] tests/crashes/120241-2.rs stdout ----

error: test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit.
thread '[crashes] tests/crashes/120241-2.rs' panicked at src/tools/compiletest/src/runtest.rs:385:18:
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- [crashes] tests/crashes/120241.rs stdout ----


error: test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit.
thread '[crashes] tests/crashes/120241.rs' panicked at src/tools/compiletest/src/runtest.rs:385:18:

---- [crashes] tests/crashes/120482.rs stdout ----


error: test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit.
thread '[crashes] tests/crashes/120482.rs' panicked at src/tools/compiletest/src/runtest.rs:385:18:

---- [crashes] tests/crashes/125512.rs stdout ----


error: test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit.
thread '[crashes] tests/crashes/125512.rs' panicked at src/tools/compiletest/src/runtest.rs:385:18:


failures:
    [crashes] tests/crashes/120241-2.rs
```

</details>

estebank committed:
Emit method not found error for object safe traits without `dyn`

Given `Trait::foo()` in edition 2021 where `Trait` is object safe and `foo` isn't resolved, emit E0782 instead of marking the expression as `{type error}`, as `foo` not being part of `Trait` is more useful information.

If `Trait` *isn't* object safe, then the resolve error isn't as useful because E0782 will have more actionable feedback.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9455598514/job/26045744374) [(plain)](https://github.com/rust-lang/rust/commit/8140dcd4b8a36129e865361b25f6a9bd19ffff07/checks/26045744374/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 28.7s done
#16 DONE 33.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[endgroup]
##[group]Testing stage2 compiletest suite=crashes mode=crashes (x86_64-unknown-linux-gnu)

running 230 tests
.............................................................................2024-06-10T21:29:22.647300Z ERROR compiletest::runtest: fatal error, panic: "test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit."
[crashes] tests/crashes/120241-2.rs ... F
[crashes] tests/crashes/120241-2.rs ... F
2024-06-10T21:29:22.654616Z ERROR compiletest::runtest: fatal error, panic: "test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit."
[crashes] tests/crashes/120241.rs ... F
[crashes] tests/crashes/120241.rs ... F
2024-06-10T21:29:22.661585Z ERROR compiletest::runtest: fatal error, panic: "test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit."
[crashes] tests/crashes/120482.rs ... F
........  88/230
........................................................................................ 176/230
........................................................................................ 176/230
..2024-06-10T21:29:23.409284Z ERROR compiletest::runtest: fatal error, panic: "test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit."
[crashes] tests/crashes/125512.rs ... F
...................................................

failures:
failures:

---- [crashes] tests/crashes/120241-2.rs stdout ----

error: test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit.
thread '[crashes] tests/crashes/120241-2.rs' panicked at src/tools/compiletest/src/runtest.rs:385:18:
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- [crashes] tests/crashes/120241.rs stdout ----


error: test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit.
thread '[crashes] tests/crashes/120241.rs' panicked at src/tools/compiletest/src/runtest.rs:385:18:

---- [crashes] tests/crashes/120482.rs stdout ----


error: test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit.
thread '[crashes] tests/crashes/120482.rs' panicked at src/tools/compiletest/src/runtest.rs:385:18:

---- [crashes] tests/crashes/125512.rs stdout ----


error: test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit.
thread '[crashes] tests/crashes/125512.rs' panicked at src/tools/compiletest/src/runtest.rs:385:18:


failures:
    [crashes] tests/crashes/120241-2.rs
```

</details>

estebank committed:
Move tests that no longer ICE

estebank commented on tests/ui/traits/bound/not-on-bare-trait-2021.stderr:
```diff
@@ -69,9 +39,13 @@ help: alternatively, use a trait object to accept any type that implements `Foo`
    |
 LL | fn bar(x: &dyn Foo) -> Foo {
    |           ++++
+help: add `dyn` keyword before this trait
+   |
+LL | fn bar(x: dyn Foo) -> Foo {
+   |           +++
```
This is wrong. I'll look into it.

estebank committed:
Do not produce incorrect `dyn Trait` suggestion in `fn`

```
  error[E0782]: trait objects must include the `dyn` keyword
    --> $DIR/not-on-bare-trait-2021.rs:8:12
     |
  LL | fn foo(_x: Foo + Send) {
     |            ^^^^^^^^^^
     |
  help: use a new generic type parameter, constrained by `Foo + Send`
     |
  LL | fn foo<T: Foo + Send>(_x: T) {
     |       +++++++++++++++     ~
  help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
     |
  LL | fn foo(_x: impl Foo + Send) {
     |            ++++
  help: alternatively, use a trait object to accept any type that implements `Foo + Send`, accessing its methods at runtime using dynamic dispatch
     |
  LL | fn foo(_x: &(dyn Foo + Send)) {
     |            +++++           +
- help: add `dyn` keyword before this trait
-    |
- LL | fn foo(_x: dyn Foo + Send) {
-    |            +++
```

estebank commented on tests/ui/suggestions/issue-116434-2021.stderr:
```diff
@@ -1,34 +1,25 @@
+   |                 ^^^^^
    |
-   = note: the trait cannot be made into an object because it requires `Self: Sized`
-   = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
-help: there is an associated type with the same name
+help: `Clone` is not object safe, use `impl Clone` to return an opaque type, as long as you return a single underlying type
    |
-LL |     fn foo() -> Self::Clone;
-   |                 ++++++
```
This is arguably a regression. I'll look into it :-/

estebank replied:
Handled in the last commit

estebank committed:
Mention associated type with same name as trait in `E0782`

```
error[E0782]: trait objects must include the `dyn` keyword
  --> $DIR/issue-116434-2021.rs:5:17
   |
LL |     type Clone;
   |     ---------- you might have meant to use this associated type
LL |     fn foo() -> Clone;
   |                 ^^^^^
   |
help: `Clone` is not object safe, use `impl Clone` to return an opaque type, as long as you return a single underlying type
   |
LL |     fn foo() -> impl Clone;
   |                 ++++
help: there is an associated type with the same name
   |
LL |     fn foo() -> Self::Clone;
   |                 ++++++
```

estebank committed:
Account for `static` and `const` in suggestion

When encountering `static FOO: Trait = ...`, if `Trait` is object safe, suggest `static FOO: Box<dyn Trait> = ...`.

estebank commented on tests/ui/const-generics/generic_const_exprs/expected-type-of-closure-body-to-be-a-closure-or-coroutine-ice-113776.stderr:
```diff
@@ -62,7 +39,7 @@ help: you might have meant to write a bound here
 LL | ) -> impl Iterator<Item: SubAssign> {
    |                        ~
```
Pre-existing, I feel that if we wait long enough to fix it, the syntax will be valid in stable 😈

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9470844569/job/26092686417) [(plain)](https://github.com/rust-lang/rust/commit/88ccfa4eb0eb4a7842cc7d479ffd8457d914d23c/checks/26092686417/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:a5ac7e51b41094c92402da3b24376905380afc29)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.422   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.435 Collecting boolean-py==4.0
#13 0.440   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.453 Collecting chardet==5.1.0
---
#13 3.555 Building wheels for collected packages: reuse
#13 3.557   Building wheel for reuse (pyproject.toml): started
#13 3.890   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.891   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.891   Stored in directory: /tmp/pip-ephem-wheel-cache-3s6gnpf3/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.893 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.914   Attempting uninstall: setuptools
#13 3.915     Found existing installation: setuptools 59.6.0
#13 3.916     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
#13 5.128   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.183      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 74.2 MB/s eta 0:00:00
#13 5.234 Collecting filelock<4,>=3.12.2
#13 5.238   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.253 Collecting distlib<1,>=0.3.7
#13 5.257   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.263      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 120.2 MB/s eta 0:00:00
#13 5.291 Collecting platformdirs<5,>=3.9.1
#13 5.296   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.382 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.541 Successfully installed distlib-0.3.8 filelock-3.14.0 platformdirs-4.2.2 virtualenv-20.26.2
#13 DONE 5.6s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      182208 kB
DirectMap2M:     8206336 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/3ea5e236ecb4c5f22437059f82d3915d311e4ec0/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-3ea5e236ecb4c5f22437059f82d3915d311e4ec0-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5052 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/tests/ui/consts/const_refs_to_static-ice-121413.rs:1: line longer than 100 chars
##[error]tidy error: /checkout/tests/ui/consts/const_refs_to_static-ice-121413.rs:2: line longer than 100 chars
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.0)
some tidy checks failed
Build completed unsuccessfully in 0:01:09
  local time: Tue Jun 11 18:30:56 UTC 2024
  network time: Tue, 11 Jun 2024 18:30:56 GMT
```

</details>

estebank committed:
Use verbose format for fully-qualified path suggestion

estebank committed:
Fix tidy

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9491311410/job/26156472346) [(plain)](https://github.com/rust-lang/rust/commit/fa382d6209f4e920799fb1197318e2ee19b3fe99/checks/26156472346/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

     Running tests/compile-test.rs (obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/compile_test-977de26160391523)

FAILED TEST: tests/ui/crashes/ice-6252.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-3bda5d1e7fb82f59.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-67be2c756df915af.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-b9bc125d38478b96.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-56069c5040049117.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-ae846d6788d6a9f1.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-e9aa61e1e4ee057e.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-3741c442f64eecc7.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-40a52e569c1b5965.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-caed1d72b1950c44.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-bf710ab22080fa03.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libserde_derive-bac580dbc61ea613.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-13e779e479165758.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-d534f89567b59ccc.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui/crashes" "tests/ui/crashes/ice-6252.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/crashes/ice-6252.stderr` to the actual output
--- tests/ui/crashes/ice-6252.stderr
+++ <stderr output>
---
full stderr:
error[E0412]: cannot find type `PhantomData` in this scope
##[error]  --> tests/ui/crashes/ice-6252.rs:9:9
   |
LL |     _n: PhantomData,
   |
help: consider importing one of these items
   |
LL + use std::marker::PhantomData;
LL + use std::marker::PhantomData;
   |
LL + use core::marker::PhantomData;
   |

error[E0412]: cannot find type `VAL` in this scope
##[error]  --> tests/ui/crashes/ice-6252.rs:11:63
   |
LL | impl<N, M> TypeVal<usize> for Multiply<N, M> where N: TypeVal<VAL> {}
   |
help: you might be missing a type parameter
   |
   |
LL | impl<N, M, VAL> TypeVal<usize> for Multiply<N, M> where N: TypeVal<VAL> {}

error[E0046]: not all trait items implemented, missing: `VAL`
##[error]  --> tests/ui/crashes/ice-6252.rs:11:1
   |
   |
LL |     const VAL: T;
   |     ------------ `VAL` from trait
...
LL | impl<N, M> TypeVal<usize> for Multiply<N, M> where N: TypeVal<VAL> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `VAL` in implementation
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0046, E0412.
For more information about an error, try `rustc --explain E0046`.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #126439) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
